---
title: "Figure 1"
author: "Benjamin Beresford-Jones"
date: "04/09/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(readr)
library(ggplot2)
library(ggpubr)
library(reshape2)
library(plyr)
library(RColorBrewer)
library(grid)
library(gplots)
```


### Set defaults
```{r}
# out path for Rplots
indir <- "~/Documents/PhD/Manuscripts/MCGC_18075/Figure_1/data/"
outdir <- "~/Documents/PhD/Manuscripts/MCGC_18075/Figure_1/"
```

```{r}
# colour palettes
#phylum_palette <- c(Actinobacteriota="#AED6F1",
#                    Bacteroidota="#b3de69",
##                    Campylobacterota="#CA87B7",
#                    Cyanobacteria="#FFFF08", #"#ffed6f", 
#                    Deferribacterota="#F9F984",
#                    Desulfobacterota="#F8B819",
#                    Elusimicrobiota="#00ffff", #"#8dd3c7", #
#                    Firmicutes="#FFC2B8",
#                    Firmicutes_A="#fb8072",
#                    Firmicutes_B="#c43b3b",
#                    Firmicutes_C="#882210",
#                    Proteobacteria="#8917EF", #"#bc80bd",  #
#                    Spirochaetota="#1AA120",
#                    Thermotogota="#3FF6B4", #"#ccebc5", #
#                    Verrucomicrobiota="#E6CAE8")

#phylum_palette <- c(Actinobacteriota="#2196F3", Bacteroidota="#4CAF50", Campylobacterota="#009688", Cyanobacteria="#80DEEA", Deferribacterota="#CDDC39", Desulfobacterota_A="#FFC107", Firmicutes="#EF5350", Firmicutes_A="#F44336", Firmicutes_B="#E53935", Firmicutes_I="#D32F2F", Patescibacteria="#F06292", Proteobacteria="#9C27B0", Verrucomicrobiota="#BCAAA4")

class_palette <- c(Actinobacteria="#80b1d3",
                   Alphaproteobacteria="#ccebc5",
                   Bacilli_A="#d9d9d9",
                   Bacilli="#ffed6f",
                   Bacteroidia="#b3de69",
                   Campylobacteria="#8dd3c7",
                   Clostridia="#fb8072",
                   Coriobacteriia="#AED6F1",
                   Deferribacteres="#ffffb3",
                   Desulfovibrionia="#fdb462",
                   Gammaproteobacteria="#bc80bd",
                   Vampirovibrionia="#E6B0AA",
                   Verrucomicrobiae="#bebada")

```

```{r}
# colour palettes
phylum_palette <- c(Actinobacteriota="#80b1d3", #"#AED6F1",
                    Bacteroidota="#b3de69",
                    Campylobacterota="#8dd3c7", #"#CA87B7",
                    Cyanobacteria="#ffe16f", # "#ffdc56", ##ffe16f", #"#ffed6f", 
                    Deferribacterota="#f9ffb3", #"#ffffb3", #"#F9F984",
                    Desulfobacterota="#fdb462", #"#F8B819",
                    Elusimicrobiota="#AED6F1", #"#8dd3c7", 
                    Firmicutes="#FFC2B8",
                    Firmicutes_A="#fb8072",
                    Firmicutes_B="#c43b3b",
                    Firmicutes_C="#882210",
                    Proteobacteria="#bc80bd",  
                    Spirochaetota="#1aa142", #"#1AA120",
                    Thermotogota="#ccebc5", 
                    Verrucomicrobiota="#E6CAE8")

# jco palette
col_HUMAN <- "#0073C2FF" # blue
#MOOUSE <- "#EFC000FF" # yellow
# #efb800 is a shade darker
col_MOUSE <- "#efb800"
```


### Get MCGC data

```{r}
MCGC <- read_tsv(file = paste0(indir,"gtdbtk_18075.tsv"), col_names = F)

colnames(MCGC) <- c("Genome", "Lowest_tax_1", "Lowest_tax_2", "Tax_rank", "Tax_name", "Taxonomy")

MCGC$Phylum <- unlist(lapply(strsplit(MCGC$Taxonomy, split = ";"), function(x) {
  gsub(x[2], pattern = "p__", replacement = "") # phylum
}))

MCGC$Class <- unlist(lapply(strsplit(MCGC$Taxonomy, split = ";"), function(x) {
  gsub(x[3], pattern = "c__", replacement = "") # class
}))

MCC_276 <- MCGC[grep(MCGC$Genome, pattern = "#"),]
MAGs <- MCGC[grep(MCGC$Genome, pattern = "#", invert = TRUE),]
Public_274 <- MAGs[grep(MAGs$Genome, pattern = "\\.", invert = TRUE),]
MAGs <- MAGs[grep(MAGs$Genome, pattern = "\\.", invert = FALSE),]
```


### Get metagenomic data

```{r}
kraken <- read_tsv(file = paste0(indir, "kraken_merge.tsv"), col_names = TRUE)

kraken$Frac_classified <- 100 - kraken$Frac_unclassified
```



## Figure 1b

```{r}
list <- lapply(split(MCGC, f = MCGC$Phylum), function(x) {
  FRAC <- nrow(x)/nrow(MCGC)*100
  NOVEL <- length(which(x$Tax_rank != "species"))/nrow(x)*100
  data.frame(FRAC, NOVEL)
})

novel_stats <- do.call("rbind", list)
novel_stats$Phylum <- row.names(novel_stats)

order <- novel_stats$Phylum[order(novel_stats$FRAC, decreasing = TRUE)]

p <- ggbarplot(novel_stats, x = "Phylum", y = "FRAC", 
               order = rev(order),
               color = "Phylum", fill = "Phylum",
               palette = phylum_palette,
               ylab = "Fraction of MCGC genomes (%)") +
  theme_pubr()
ggpar(p, legend = "right", rotate = TRUE, ylab = FALSE)

p <- ggbarplot(novel_stats, x = "Phylum", y = "NOVEL", 
               order = rev(order),
               color = "Phylum", fill = "Phylum",
               palette = phylum_palette,
               ylab = "Fraction of uncharacterised genomes by phylum") +
  theme_pubr()
ggpar(p, legend = "right", rotate = TRUE, ylab = FALSE)
```

```{r}
novel_stats$Phylum <- factor(novel_stats$Phylum, levels = rev(order))

g.mid<-ggplot(novel_stats,aes(x=1,y=Phylum)) + geom_text(aes(label=Phylum))+
  geom_segment(aes(x=0.95,xend=0.952,yend=Phylum))+
  ggtitle("")+
  ylab(NULL)+
  scale_x_continuous(expand=c(0,0),limits=c(0.95,1.05))+
  theme(axis.title=element_blank(),
        panel.grid=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank(),
        panel.background=element_blank(),
        axis.text.x=element_text(color=NA),
        axis.ticks.x=element_line(color=NA),
        plot.margin = unit(c(2,0,2,-1), "mm"), # c(2,0,2,-1)
        plot.title = element_text(hjust = 0.5)
        )


g1 <- ggplot(data = novel_stats, aes(x = Phylum, y = FRAC, color = Phylum, fill = Phylum)) +
  geom_bar(stat = "identity") + ggtitle("% MCGC genomes") +
  theme(axis.title.x = element_blank(), 
        axis.title.y = element_blank(), 
        axis.text.y = element_blank(), 
        axis.ticks.y = element_blank(),
        axis.text.x = element_text(colour = "black"),
        panel.border = element_rect(color = "black", fill = NA),
        panel.background = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.grid.minor.x = element_blank(),
        legend.position = "none",
        plot.margin = unit(c(1,0,1,1), "mm"), # c(1,0,1,1)
        plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=phylum_palette) +
  scale_fill_manual(values=phylum_palette) +
  scale_y_reverse() + 
  coord_flip()
  

g2 <- ggplot(data = novel_stats, aes(x = Phylum, y = NOVEL, color = Phylum, fill = Phylum)) +xlab(NULL)+
  geom_bar(stat = "identity") + ggtitle("% Novel genomes") +
  theme(axis.title.x = element_blank(), axis.title.y = element_blank(), 
        axis.text.y = element_blank(), 
        axis.text.x = element_text(colour = "black"),
        panel.border = element_rect(color = "black", fill = NA),
        panel.background = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.grid.minor.x = element_blank(),
        legend.position = "none",
        plot.margin = unit(c(1,1,1,-1), "mm"), # c(1,1,1,-1)
        plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=phylum_palette) +
  scale_fill_manual(values=phylum_palette) +
  coord_flip() 



library(gridExtra)
gg1 <- ggplot_gtable(ggplot_build(g1))
gg2 <- ggplot_gtable(ggplot_build(g2))
gg.mid <- ggplot_gtable(ggplot_build(g.mid))


grid.arrange(gg1,gg.mid,gg2,ncol=3,widths=c(7/18,4/18,7/18))
```

```{r}
pdf(paste0(outdir, "MMGC_phyla_novelty.pdf"), height = 3, width = 9.67)
grid.arrange(gg1,gg.mid,gg2,ncol=3,widths=c(7/18,4/18,7/18))
dev.off()
```

```{r}
novel_stats
```



## Figure 1d

```{r}
unique(kraken$Database)
```

```{r}
# remove addition kraken database analyses
kraken <- kraken[which(kraken$Database != "HUMAN-4594"),]

# relabel databases
kraken$Database[which(kraken$Database == "HUMAN_REPS-3113")] <- "Human"
kraken$Database[which(kraken$Database == "ISOLATES-550")] <- "MCC+Public"
kraken$Database[which(kraken$Database == "MCC-276")] <- "MCC"
kraken$Database[which(kraken$Database == "MCGC_REPS-1924")] <- "MMGC"
kraken$Database[which(kraken$Database == "mGMB-100")] <- "mGMB"
kraken$Database[which(kraken$Database == "miBC-47")] <- "miBC"
kraken$Database[which(kraken$Database == "PUBLIC-274")] <- "Public"

# order and addhost metadata for figure colour
kDB_order <- c("miBC", "mGMB", "Public", "MCC", "MCC+Public", "MMGC", "Human")
kraken$Host <- "Mouse"
kraken$Host[which(kraken$Database == "Human")] <- "Human"
```

### without qc

```{r}
# summarise database efficiencies
summary(kraken$Frac_classified[kraken$Database == "miBC"])
summary(kraken$Frac_classified[kraken$Database == "mGMB"])
summary(kraken$Frac_classified[kraken$Database == "Public"])
summary(kraken$Frac_classified[kraken$Database == "MCC"])
summary(kraken$Frac_classified[kraken$Database == "MCC+Public"])
summary(kraken$Frac_classified[kraken$Database == "MMGC"])
summary(kraken$Frac_classified[kraken$Database == "Human"])
```


```{r}
p <- ggboxplot(data = kraken, x = "Database", y = "Frac_classified",
              #add = "jitter", add.params = list(width = 0.8), 
              #add = "dotplot", add.params = list(binwidth = 2, size = 0.1, width = 0.8),
              color = "Host", 
              #fill = "Host",
              palette = c(Mouse = col_MOUSE, Human = col_HUMAN),
              #palette = c(Mouse = "#EFC000FF", Human = "#0073C2FF"), # JCO
              #palette = "npg",
              xlab = "Database", ylab = "Read classification (%)", 
              order = kDB_order,
              #x.text.angle = 20,
              font.y = 16,
              notch = TRUE,
              legend = "none", 
              remove = c("All"),
              outlier.shape=NA
) +
  theme(legend.position = "none")


kraken_comparisons <- list(c("Public", "MCC"),
                         c("Public", "MCC+Public"), 
                         #c("Public", "MCGC"),
                         c("MCC", "Human"))

                         

ggpar(p, xlab = FALSE) + 
  stat_compare_means(comparisons = kraken_comparisons,
                     method = "t.test", paired = TRUE, 
                     label = "p.signif", label.y = c(84,90,100)) +
  theme(axis.line = element_blank(),
    panel.grid.minor = element_blank(),
    panel.grid.major.y = element_line(colour = "lightgrey"),
    axis.title.x = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA),
    legend.position = "none",
    axis.title.y = element_text(size = 18)
  ) 
```

```{#r}
#jpeg("~/Documents/PhD/Manuscripts/Human_Mouse_comparison/Figure 1/R_plots/rplot_4.jpg", height = 375, width = 600)
ggpar(p, xlab = FALSE) + 
  stat_compare_means(comparisons = my_comparisons_3,
                     method = "t.test", paired = TRUE, 
                     label = "p.signif", label.y = c(84,90,100)) +
  theme(axis.line = element_blank(),
    panel.grid.minor = element_blank(),
    panel.grid.major.y = element_line(colour = "lightgrey"),
    axis.title.x = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA),
    legend.position = "none",
    axis.title.y = element_text(size = 18)
  ) 
dev.off()
```

```{r}
kraken_rep_qc
```



## Human kraken greater than mouse?

```{r}
kDB_mmgc <- kraken[which(kraken$Database=="MMGC"),]
kDB_human <- kraken[which(kraken$Database=="Human"),]

which(kDB_mmgc$Frac_classified - kDB_human$Frac_classified < 0)

sort(kDB_mmgc$Frac_classified - kDB_human$Frac_classified)
```


### with qc

```{r}
MAG_samples <- as.character(unlist(read_csv(file = paste0(indir, "successful_MAG_samples.txt"), col_names = FALSE)))
```

```{r}
sample_ids <- unlist(lapply(strsplit(unique(kraken$Sample), split = "_"), function(x) { 
  gsub(pattern = "\\.3", replacement = "", x[1]) 
  }))

# add sample ids for mgp15975 samples (all of which produced MAGs)
MAG_samples <- c(MAG_samples, 
  sample_ids[grep("^C.$|^E.$|^T.$", sample_ids)])
```

```{r}
qc_samples <- unique(kraken$Sample)[which(sample_ids %in% MAG_samples)]
qc_kraken <- kraken[which(kraken$Sample %in% qc_samples),]
```

```{r}
# summarise database efficiencies
summary(qc_kraken$Frac_classified[qc_kraken$Database == "miBC"])
summary(qc_kraken$Frac_classified[qc_kraken$Database == "mGMB"])
summary(qc_kraken$Frac_classified[qc_kraken$Database == "Public"])
summary(qc_kraken$Frac_classified[qc_kraken$Database == "MCC"])
summary(qc_kraken$Frac_classified[qc_kraken$Database == "MCC+Public"])
summary(qc_kraken$Frac_classified[qc_kraken$Database == "MMGC"])
summary(qc_kraken$Frac_classified[qc_kraken$Database == "Human"])
```


```{r}
p <- ggboxplot(data = qc_kraken, x = "Database", y = "Frac_classified",
              #add = "jitter", add.params = list(width = 0.8), 
              #add = "dotplot", add.params = list(binwidth = 2, size = 0.1, width = 0.8),
              color = "Host", 
              #fill = "Host",
              palette = c(Mouse = col_MOUSE, Human = col_HUMAN),
              #palette = c(Mouse = "#EFC000FF", Human = "#0073C2FF"), # JCO
              #palette = "npg",
              xlab = "Database", ylab = "Read classification (%)", 
              order = kDB_order,
              #x.text.angle = 20,
              font.y = 16,
              notch = TRUE,
              legend = "none", 
              remove = c("All"),
              outlier.shape=NA
) +
  theme(legend.position = "none")


qc_kraken_comparisons <- list(c("Public", "MCC"),
                         c("Public", "MCC+Public"),
                         #c("Public", "MCGC"),
                         c("MCC", "Human"))

                         

ggpar(p, xlab = FALSE) + 
  stat_compare_means(comparisons = qc_kraken_comparisons,
                     method = "t.test", paired = TRUE, 
                     label = "p.signif", label.y = c(84,90,100)) +
  theme(axis.line = element_blank(),
    panel.grid.minor = element_blank(),
    panel.grid.major.y = element_line(colour = "lightgrey"),
    axis.title.x = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA),
    legend.position = "none",
    axis.title.y = element_text(size = 18)
  ) 
```

```{r}
p <- ggboxplot(data = qc_kraken, x = "Database", y = "Frac_classified",
              color = "Host", 
              palette = c(Mouse = col_MOUSE, Human = col_HUMAN),
              xlab = "Database", ylab = "Read classification (%)", 
              order = kDB_order,
              font.y = 16,
              notch = TRUE,
              #legend = "none", 
              remove = c("All"),
              outlier.shape=NA
) +
  theme(legend.position = c(0.15,0.78), 
        legend.title = element_blank())


qc_kraken_comparisons <- list(c("Public", "MCC"),
                         c("Public", "MCC+Public"), 
                         c("MCC+Public", "MMGC"),
                         c("MCC", "Human"))


ggpar(p, xlab = FALSE) + 
  stat_compare_means(comparisons = qc_kraken_comparisons,
                     method = "t.test", paired = TRUE, 
                     label = "p.signif", label.y = c(84,92.5,101,109.5)) +
  theme(#axis.line = element_blank(),
    panel.grid.minor = element_blank(),
    panel.grid.major.y = element_line(colour = "lightgrey"),
    axis.title.x = element_blank(),
    #panel.border = element_rect(colour = "black", fill = NA),
    #legend.position = "none",
    axis.title.y = element_text(size = 18)
  ) +
  scale_y_continuous(breaks=seq(0,100,25))
```


## Remove samples with derived MMGC reps

```{r}
rep_samples <- as.character(unlist(read_csv(file = paste0(indir, "MCGC_rep_samples.txt"), col_names = FALSE)))
```


```{r}
# get non-rep-contaiing samples
non_rep_samples <- unique(kraken$Sample)[which(!(sample_ids %in% rep_samples))]
# get the post-qc non-rep-containing samples
rep_qc_kraken <- qc_kraken[which(qc_kraken$Sample %in% non_rep_samples),]
```

```{r}
length(unique(qc_kraken$Sample))
length(unique(rep_qc_kraken$Sample))
```


```{r}
# summarise database efficiencies
summary(rep_qc_kraken$Frac_classified[rep_qc_kraken$Database == "miBC"])
summary(rep_qc_kraken$Frac_classified[rep_qc_kraken$Database == "mGMB"])
summary(rep_qc_kraken$Frac_classified[rep_qc_kraken$Database == "Public"])
summary(rep_qc_kraken$Frac_classified[rep_qc_kraken$Database == "MCC"])
summary(rep_qc_kraken$Frac_classified[rep_qc_kraken$Database == "MCC+Public"])
summary(rep_qc_kraken$Frac_classified[rep_qc_kraken$Database == "MMGC"])
summary(rep_qc_kraken$Frac_classified[rep_qc_kraken$Database == "Human"])
```


```{r}
p <- ggboxplot(data = rep_qc_kraken, x = "Database", y = "Frac_classified",
              fill = "Host", color = "black",
              palette = c(Mouse = col_MOUSE, Human = col_HUMAN),
              xlab = "Database", ylab = "Read classification (%)", 
              order = kDB_order,
              font.y = 16,
              notch = TRUE,
              #legend = "none", 
              remove = c("All"),
              outlier.shape=NA
) +
  theme(legend.position = c(0.15,0.765), 
        legend.title = element_blank())


rep_qc_kraken_comparisons <- list(#c("Public", "MCC"),
                         c("Public", "MCC+Public"), 
                         c("MCC+Public", "MMGC"),
                         c("MCC", "Human")
                         )


ggpar(p, xlab = FALSE) + 
  stat_compare_means(comparisons = rep_qc_kraken_comparisons,
                     method = "t.test", paired = TRUE, 
                     label = "p.signif", label.y = c(#87,
                                                     95,103,111)) +
  theme(#axis.line = element_blank(),
    panel.grid.minor = element_blank(),
    #panel.grid.major.y = element_line(colour = "lightgrey"),
    axis.title.x = element_blank(),
    #panel.border = element_rect(colour = "black", fill = NA),
    #legend.position = "none",
    axis.title.y = element_text(size = 18)
  ) +
  scale_y_continuous(breaks=seq(0,100,25))
```


```{r}
pdf(paste0(outdir, "kraken_improvement_woGrid.pdf"), height = 4.5, width = 8)
ggpar(p, xlab = FALSE) + 
  stat_compare_means(comparisons = rep_qc_kraken_comparisons,
                     method = "t.test", paired = TRUE, 
                     label = "p.signif", label.y = c(#87,
                                                     95,103,111)) +
  theme(#axis.line = element_blank(),
    panel.grid.minor = element_blank(),
    #panel.grid.major.y = element_line(colour = "lightgrey"),
    axis.title.x = element_blank(),
    #panel.border = element_rect(colour = "black", fill = NA),
    #legend.position = "none",
    axis.title.y = element_text(size = 18)
  ) +
  scale_y_continuous(breaks=seq(0,100,25))
dev.off()
```


```{r}
rep_qc_kraken
```



## Figure 1c

```{r}
p_bracken <- read_tsv(file = paste0(indir, "merged_bracken.P.tsv"), col_names = TRUE)
```


### Remove qc'ed samples from brazcken analyses
```{r}
qc_names <- gsub(pattern = "\\.out", replacement = "", qc_samples)

qc_bracken <- p_bracken[which(p_bracken$Sample %in% qc_names),]
qc_bracken$Species_read_fraction <- qc_bracken$Species_read_fraction*100 
qc_bracken$Species[which(qc_bracken$Species == "Desulfobacterota_A")] <- "Desulfobacterota"

# remove misincluded Firmicutes_I
phylum_qc <- qc_bracken[which(!qc_bracken$Species == "Firmicutes_I"),]
```


## Average microbiome
```{r}
p <- ggbarplot(data = phylum_qc,
          x = "Sample", y = "Species_read_fraction",
          ylab = "Proportion of classified reads (%)",
          xlab = "Sample",
          color = "Species", fill = "Species",
          palette = phylum_palette
          ) + 
  theme_pubclean()
ggpar(p, legend = "right",
      font.xtickslab = FALSE, 
      ticks = FALSE)
```

```{r}
microbiome <- data.frame(Read_frac=unlist(lapply(split(phylum_qc, f = phylum_qc$Species), function(x) {
  mean(x$Species_read_fraction)
}
  )))

microbiome$Phylum <- row.names(microbiome)
microbiome$Collection <- "Microbiome"

p <- ggbarplot(microbiome, y = "Read_frac", color = "Phylum", fill = "Phylum",
          palette = phylum_palette)
ggpar(p, legend = "right")
```

# Get MCGC/MCC breakdown
## MCGC
```{r}
MCGC_phylum <- data.frame(Read_frac=unlist(lapply(split(MCGC, f = MCGC$Phylum), function(x) {
  nrow(x)/nrow(MCGC)*100
})))

MCGC_phylum$Phylum <- row.names(MCGC_phylum)
MCGC_phylum$Collection <- "MMGC"

p <- ggbarplot(MCGC_phylum, y = "Read_frac", color = "Phylum", fill = "Phylum",
               palette = phylum_palette)
ggpar(p, legend = "right")
```

## MCC
```{r}
MCC_phylum <- data.frame(Read_frac=unlist(lapply(split(MCC_276, f = MCC_276$Phylum), function(x) {
  nrow(x)/nrow(MCC_276)*100
})))

MCC_phylum$Phylum <- row.names(MCC_phylum)
MCC_phylum$Collection <- "MCC isolates"

p <- ggbarplot(MCC_phylum, y = "Read_frac", color = "Phylum", fill = "Phylum",
          palette = phylum_palette)
ggpar(p, legend = "right")
```

```{r}
collection_comparison <- rbind(MCC_phylum, MCGC_phylum, microbiome)

p <- ggbarplot(data = collection_comparison, 
          x = "Collection", y = "Read_frac",
          ylab = "Proportion of classified reads (%)",
          color = "Phylum", fill = "Phylum",
          palette = phylum_palette,
           order = c("MCC isolates", "MMGC", "Microbiome"))

ggpar(p, legend = "none", xlab = FALSE) +
  theme(panel.grid.minor = element_blank(),
    panel.grid.major.y = element_blank(),
    axis.title.x = element_blank(),
    legend.position = "none",
    axis.title.y = element_text(size = 18)
  ) 
```

```{r}
pdf(paste0(outdir, "MMGC_microbiome_representation.pdf"), height = 5, width = 5)
ggpar(p, legend = "none", xlab = FALSE) +
  theme(panel.grid.minor = element_blank(),
    panel.grid.major.y = element_blank(),
    axis.title.x = element_blank(),
    legend.position = "none",
    axis.title.y = element_text(size = 18)
  ) 
dev.off()
```


```{r}
tmpPhylum <- data.frame(Phylum=MCGC_phylum$Phylum)

tmpMCGC <- merge(x=tmpPhylum, y=MCGC_phylum[,c(2,1)], by="Phylum", all.x = TRUE)
colnames(tmpMCGC)[2] <- c("MCGC")      

tmpMCC <- merge(x=tmpPhylum, y=MCC_phylum[,c(2,1)], by="Phylum", all.x = TRUE)
colnames(tmpMCC)[2] <- c("MCC")      
      
tmpMB <- merge(x=tmpPhylum, y=microbiome[,c(2,1)], by="Phylum", all.x = TRUE)
colnames(tmpMB)[2] <- c("Microbiome")

tmpColCom <- merge(tmpMCGC, tmpMCC, by = "Phylum")
tmpColCom <- merge(tmpColCom, tmpMB, by = "Phylum")

tmpColCom[is.na(tmpColCom)] <- 0

colnames(tmpColCom)
```

Determine statistics for comparison.
```{r}
dt <- as.table(as.matrix(tmpColCom[,c(2:4)]))
row.names(dt) <- as.character(unlist(tmpPhylum))

## MCGC vs MB
chisq.test(tmpColCom[,c(2,4)])
# p-value = 0.9999


## MCC vs MB
chisq.test(tmpColCom[,c(3,4)])
# p-value = 0.01334

```
